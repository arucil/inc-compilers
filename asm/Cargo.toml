[package]
name = "asm"
version = "0.1.0"
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
indexmap = "1.8.1"
enum-iterator = "0.8.1"
num-derive = "0.3.3"
num-traits = "0.2.15"
support = { path = "../support" }
