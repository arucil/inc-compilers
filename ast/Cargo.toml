[package]
name = "ast"
version = "0.1.0"
edition = "2018"

[dependencies]
pretty = "0.10.0"
support = { path = "../support" }

[dev-dependencies]
pretty_assertions = "0.7.2"
