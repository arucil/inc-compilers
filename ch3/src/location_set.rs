use asm::{Arg, Reg};
use ast::IdxVar;
use indexmap::map::Entry;
use indexmap::IndexMap;
use num_traits::{FromPrimitive, ToPrimitive};
use smallvec::SmallVec;
use std::fmt::{self, Debug, Write};
use std::ops::Deref;

#[derive(Debug, Clone)]
pub struct LocationSet(SmallVec<[u32; 2]>);

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct Location(usize);

#[derive(Debug, Clone)]
pub struct VarStore(IndexMap<IdxVar, Var>);

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct Var(usize);

impl LocationSet {
  pub fn new() -> Self {
    Self(SmallVec::new())
  }

  pub fn add_reg(&mut self, reg: Reg) {
    let i = reg_index(reg);
    if self.0.is_empty() {
      self.0.push(0);
    }
    self.0[i / 32] |= 1 << (i % 32);
  }

  pub fn remove_reg(&mut self, reg: Reg) {
    let i = reg_index(reg);
    if !self.0.is_empty() {
      self.0[i / 32] &= !(1 << (i % 32));
    }
  }

  pub fn add_var(&mut self, var: Var) {
    let i = var_index(var);
    let offset = i / 32;
    if offset >= self.0.len() {
      self.0.resize(offset + 1, 0);
    }
    self.0[offset] |= 1 << (i % 32);
  }

  pub fn remove_var(&mut self, var: Var) {
    let i = var_index(var);
    let offset = i / 32;
    if offset < self.0.len() {
      self.0[offset] &= !(1 << (i % 32));
    }
  }

  pub fn remove_caller_saved_regs(&mut self) {
    self.remove_reg(Reg::Rax);
    self.remove_reg(Reg::Rcx);
    self.remove_reg(Reg::Rdx);
    self.remove_reg(Reg::Rsi);
    self.remove_reg(Reg::Rdi);
    self.remove_reg(Reg::R8);
    self.remove_reg(Reg::R9);
    self.remove_reg(Reg::R10);
    self.remove_reg(Reg::R11);
  }

  pub fn add_argument_regs(&mut self, arity: usize) {
    const ARGUMENT_REGS: [Reg; 6] =
      [Reg::Rdi, Reg::Rsi, Reg::Rdx, Reg::Rcx, Reg::R8, Reg::R9];

    assert!(arity <= 6);
    for reg in &ARGUMENT_REGS[..arity] {
      self.add_reg(*reg);
    }
  }

  pub fn iter(&self) -> LocationIter {
    let current = if self.0.len() > 0 { self.0[0] } else { 0 };
    LocationIter {
      set: self,
      index: 0,
      current,
    }
  }

  pub fn write(&self, f: &mut impl Write, var_store: &VarStore) -> fmt::Result {
    write!(f, "{{")?;
    let mut comma = false;
    for elem in self {
      if comma {
        write!(f, ", ")?;
      }
      comma = true;
      write!(f, "{:?}", elem.to_arg(var_store))?;
    }
    write!(f, "}}")
  }
}

pub struct LocationIter<'a> {
  set: &'a LocationSet,
  index: usize,
  current: u32,
}

impl<'a> Iterator for LocationIter<'a> {
  type Item = Location;

  fn next(&mut self) -> Option<Location> {
    while self.current == 0 {
      self.index += 1;
      if self.index >= self.set.0.len() {
        return None;
      }
      self.current = self.set.0[self.index];
    }
    let i = self.current.trailing_zeros() as usize + self.index * 32;
    self.current &= self.current - 1;
    Some(Location(i))
  }
}

impl<'a> IntoIterator for &'a LocationSet {
  type Item = Location;
  type IntoIter = LocationIter<'a>;

  fn into_iter(self) -> Self::IntoIter {
    self.iter()
  }
}

fn reg_index(reg: Reg) -> usize {
  reg.to_usize().unwrap()
}

fn var_index(var: Var) -> usize {
  var.0 + 16
}

impl VarStore {
  pub fn new() -> Self {
    Self(IndexMap::new())
  }

  pub fn get(&mut self, var: IdxVar) -> Var {
    let len = self.0.len();
    match self.0.entry(var) {
      Entry::Occupied(var) => *var.get(),
      Entry::Vacant(entry) => *entry.insert(Var(len)),
    }
  }
}

impl Deref for VarStore {
  type Target = IndexMap<IdxVar, Var>;

  fn deref(&self) -> &Self::Target {
    &self.0
  }
}

impl Var {
  pub fn to_arg(&self, var_store: &VarStore) -> Arg<IdxVar> {
    Arg::Var(var_store.get_index(self.0 - 16).unwrap().0.clone())
  }
}

impl Location {
  pub fn to_arg(&self, var_store: &VarStore) -> Arg<IdxVar> {
    if self.0 < 16 {
      Arg::Reg(Reg::from_usize(self.0).unwrap())
    } else {
      Arg::Var(var_store.0.get_index(self.0 - 16).unwrap().0.clone())
    }
  }

  pub fn from_arg(arg: Arg<IdxVar>, var_store: &mut VarStore) -> Option<Self> {
    match arg {
      Arg::Deref(reg, _) | Arg::Reg(reg) => Some(reg.into()),
      Arg::Var(var) => Some(Self(var_store.get(var).0 + 16)),
      Arg::Imm(_) => None,
      Arg::ByteReg(_) => unimplemented!(),
    }
  }

  pub fn to_reg(&self) -> Option<Reg> {
    Reg::from_usize(self.0)
  }

  pub fn to_var(&self) -> Option<Var> {
    if self.0 >= 16 {
      Some(Var(self.0 - 16))
    } else {
      None
    }
  }
}

impl From<Reg> for Location {
  fn from(reg: Reg) -> Self {
    Self(reg.to_usize().unwrap())
  }
}
