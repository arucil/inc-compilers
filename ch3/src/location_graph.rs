use crate::location_set::Location;
use indexmap::IndexMap;
use petgraph::graph::{Graph, NodeIndex as PrivateNodeIndex, UnGraph};
use std::marker::PhantomData;

pub struct LocationGraph<T> {
  graph: UnGraph<Location, ()>,
  nodes: IndexMap<Location, PrivateNodeIndex>,
  _marker: PhantomData<T>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct NodeIndex<T> {
  index: PrivateNodeIndex,
  _marker: PhantomData<T>,
}

impl<T> LocationGraph<T> {
  pub fn new() -> Self {
    Self {
      graph: Graph::new_undirected(),
      nodes: IndexMap::new(),
      _marker: PhantomData,
    }
  }

  pub fn insert_node(&mut self, loc: Location) -> NodeIndex<T> {
    let index = if let Some(&node) = self.nodes.get(&loc) {
      node
    } else {
      let node = self.graph.add_node(loc);
      self.nodes.insert(loc, node);
      node
    };
    NodeIndex {
      index,
      _marker: PhantomData,
    }
  }

  pub fn node_index(&self, loc: Location) -> Option<NodeIndex<T>> {
    self.nodes.get(&loc).map(|&index| NodeIndex {
      index,
      _marker: PhantomData,
    })
  }

  pub fn node_data(&self, node: NodeIndex<T>) -> Option<Location> {
    self.graph.node_weight(node.index).cloned()
  }

  pub fn add_edge(&mut self, n1: NodeIndex<T>, n2: NodeIndex<T>) {
    if !self.graph.contains_edge(n1.index, n2.index) {
      self.graph.add_edge(n1.index, n2.index, ());
    }
  }

  pub fn node_indices(&self) -> impl Iterator<Item = NodeIndex<T>> {
    self.graph.node_indices().map(|index| NodeIndex {
      index,
      _marker: PhantomData,
    })
  }

  pub fn neighbors(
    &self,
    node: NodeIndex<T>,
  ) -> impl Iterator<Item = NodeIndex<T>> + '_ {
    self.graph.neighbors(node.index).map(|index| NodeIndex {
      index,
      _marker: PhantomData,
    })
  }

  pub fn graph(&self) -> &UnGraph<Location, ()> {
    &self.graph
  }
}
