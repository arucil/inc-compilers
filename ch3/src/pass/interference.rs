use super::liveness_analysis::Info as OldInfo;
use crate::location_graph::LocationGraph;
use crate::location_set::{Location, LocationSet, VarStore};
use asm::{Block, Instr, Program, Reg};
use ast::IdxVar;
use indexmap::IndexSet;
use petgraph::dot::{Config, Dot};
use std::fmt::{self, Debug, Formatter};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct Interference;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct Moves;

pub struct Info {
  /// Includes all locals.
  pub locals: IndexSet<IdxVar>,
  /// Nodes include all locals.
  pub conflicts: LocationGraph<Interference>,
  pub moves: LocationGraph<Moves>,
  pub var_store: VarStore,
}

pub fn build_interference(
  prog: Program<OldInfo, IdxVar>,
) -> Program<Info, IdxVar> {
  let var_store = &prog.info.var_store;
  let live = &prog.info.live;
  let mut conflicts = LocationGraph::new();
  for (_, var) in var_store.iter() {
    conflicts.insert_node(Location::from(var.clone()));
  }
  for (label, block) in &prog.blocks {
    build_graph(block, &live[label], var_store, &mut conflicts);
  }

  Program {
    info: Info {
      locals: prog.info.locals,
      conflicts,
      moves: LocationGraph::new(),
      var_store: prog.info.var_store,
    },
    constants: prog.constants,
    blocks: prog.blocks,
  }
}

fn build_graph(
  block: &Block<IdxVar>,
  live_sets: &[LocationSet],
  var_store: &VarStore,
  graph: &mut LocationGraph<Interference>,
) {
  for (instr, live_after) in block.code.iter().zip(live_sets.iter().skip(1)) {
    add_instr_edges(instr, live_after, var_store, graph);
  }
}

fn add_instr_edges(
  instr: &Instr<IdxVar>,
  live_after: &LocationSet,
  var_store: &VarStore,
  graph: &mut LocationGraph<Interference>,
) {
  let mut add = |write_loc: Location| {
    let write_loc_node = graph.insert_node(write_loc);
    for after_loc in live_after {
      if after_loc != write_loc {
        let after_loc_node = graph.insert_node(after_loc);
        graph.add_edge(write_loc_node, after_loc_node);
      }
    }
  };

  match instr {
    Instr::Add { dest, .. }
    | Instr::Sub { dest, .. }
    | Instr::Xor { dest, .. }
    | Instr::Neg(dest)
    | Instr::Pop(dest) => {
      if let Some(dest_loc) = Location::from_arg(dest.clone(), var_store) {
        add(dest_loc);
      }
    }
    Instr::Call { .. } => {
      for reg in Reg::caller_saved_regs() {
        add(reg.into());
      }
    }
    Instr::Mov { src, dest } | Instr::Movzx { src, dest } => {
      if let Some(dest_loc) = Location::from_arg(dest.clone(), var_store) {
        if let Some(src_loc) = Location::from_arg(src.clone(), var_store) {
          let dest_loc_node = graph.insert_node(dest_loc);
          for after_loc in live_after {
            if after_loc != dest_loc && after_loc != src_loc {
              let after_loc_node = graph.insert_node(after_loc);
              graph.add_edge(dest_loc_node, after_loc_node);
            }
          }
        } else {
          add(dest_loc);
        }
      }
    }
    Instr::Cmp { .. }
    | Instr::Push(_)
    | Instr::Ret
    | Instr::Syscall
    | Instr::Jmp(_)
    | Instr::JumpIf { .. } => {}
    _ => todo!(),
  }
}

impl Debug for Info {
  fn fmt(&self, f: &mut Formatter) -> fmt::Result {
    let graph = self
      .conflicts
      .graph()
      .map(|_, var| var.to_arg(&self.var_store), |_, _| ());
    Dot::with_config(&graph, &[Config::EdgeNoLabel]).fmt(f)
  }
}

#[cfg(test)]
mod tests {
  use super::super::*;
  use super::*;
  use asm::Label;
  use ch2::pass::select_instruction::Info as OldOldInfo;
  use indexmap::indexset;
  use insta::assert_snapshot;
  use maplit::hashmap;

  #[test]
  fn example_in_book() {
    use asm::Reg::*;
    let blocks = asm::parse_blocks(
      |s| IdxVar::new(s),
      r#"
    start:
      mov v, i
      mov w, 42
      mov x, v
      add x, 7
      mov y, x
      mov z, x
      add z, w
      mov t, y
      neg t
      mov rax, z
      add rax, t
      jmp conclusion
    "#,
    );
    let label_live = hashmap! {
      Label::Conclusion => LocationSet::regs([Rax, Rsp])
    };
    let prog = Program {
      info: OldOldInfo {
        locals: indexset! {
          IdxVar::new("v"),
          IdxVar::new("i"),
          IdxVar::new("w"),
          IdxVar::new("x"),
          IdxVar::new("y"),
          IdxVar::new("z"),
          IdxVar::new("t")
        },
      },
      constants: Default::default(),
      blocks,
    };
    let prog = liveness_analysis::analyze_liveness(prog, label_live);
    let result = build_interference(prog);

    assert_snapshot!(format!("{:?}", result.info));
  }

  #[test]
  fn call() {
    use asm::Reg::*;
    let blocks = asm::parse_blocks(
      |s| IdxVar::new(s),
      r#"
    start:
      pop rdi
      pop rsi
      push x
      mov w, rbx
      call foo, 3
      add w, rax
      jmp conclusion
    "#,
    );
    let label_live = hashmap! {
      Label::Conclusion => LocationSet::regs([Rax, Rsp])
    };
    let prog = Program {
      info: OldOldInfo {
        locals: indexset! {
          IdxVar::new("x"),
          IdxVar::new("w"),
        },
      },
      constants: Default::default(),
      blocks,
    };
    let prog = liveness_analysis::analyze_liveness(prog, label_live);
    let result = build_interference(prog);

    assert_snapshot!(format!("{:?}", result.info));
  }

  #[test]
  fn mov_same_variables() {
    let blocks = asm::parse_blocks(
      |s| IdxVar::new(s),
      r#"
    start:
      mov t, x
      add t, y
      mov z, t
      add z, w
    "#,
    );
    let label_live = hashmap! {};
    let prog = Program {
      info: OldOldInfo {
        locals: indexset! {
          IdxVar::new("t"),
          IdxVar::new("x"),
          IdxVar::new("y"),
          IdxVar::new("z"),
          IdxVar::new("w"),
        },
      },
      constants: Default::default(),
      blocks,
    };
    let prog = liveness_analysis::analyze_liveness(prog, label_live);
    let result = build_interference(prog);

    assert_snapshot!(format!("{:?}", result.info));
  }
}
