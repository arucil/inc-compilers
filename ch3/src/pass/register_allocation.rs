use super::interference::Info as OldInfo;
use crate::location_graph::LocationGraph;
use crate::location_set::{Location, Var, VarStore};
use asm::{Arg, Block, Instr, Program, Reg};
use ast::IdxVar;
use indexmap::{IndexMap, IndexSet};
use num_traits::ToPrimitive;
use petgraph::graph::NodeIndex;
use priority_queue::PriorityQueue;
use std::cmp::Ordering;
use std::collections::{BTreeSet, HashMap, HashSet};
use std::fmt::{self, Debug, Formatter};

pub struct Info {
  pub locals: IndexSet<IdxVar>,
  /// in bytes
  pub stack_space: usize,
  pub used_callee_saved_regs: IndexSet<Reg>,
}

impl Debug for Info {
  fn fmt(&self, f: &mut Formatter) -> fmt::Result {
    writeln!(f, "locals: {:?}", self.locals)?;
    writeln!(
      f,
      "used_callee_saved_regs: {:?}",
      self.used_callee_saved_regs
    )?;
    writeln!(f, "stack_space: {} bytes", self.stack_space)
  }
}

pub fn allocate_registers(
  prog: Program<OldInfo, IdxVar>,
  available_regs: &[Reg],
) -> Program<Info> {
  let reg_colors: HashMap<_, _> = available_regs
    .iter()
    .enumerate()
    .map(|(i, &reg)| (reg, i as i32))
    .collect();
  let mut var_store = prog.info.var_store;
  let conflicts = prog.info.conflicts;
  let moves = prog.info.moves;
  let locals: Vec<_> = prog
    .info
    .locals
    .iter()
    .map(|var| var_store.get(var.clone()))
    .collect();
  let mut max_locals = 0;
  let mut used_callee_saved_regs = IndexSet::new();
  let tmp_moves = LocationGraph::new();

  let blocks = prog
    .blocks
    .into_iter()
    .map(|(label, block)| {
      let conflicts = &conflicts[&label];
      let moves = moves.get(&label).unwrap_or(&tmp_moves);
      let (block, locals) = allocate_registers_block(
        block,
        &locals,
        &mut var_store,
        conflicts,
        moves,
        &reg_colors,
        available_regs,
        &mut used_callee_saved_regs,
      );
      if locals > max_locals {
        max_locals = locals;
      }
      (label, block)
    })
    .collect();

  Program {
    info: Info {
      locals: prog.info.locals,
      stack_space: max_locals * 8,
      used_callee_saved_regs,
    },
    blocks,
  }
}

fn allocate_registers_block(
  block: Block<IdxVar>,
  locals: &[Var],
  var_store: &mut VarStore,
  conflicts: &LocationGraph,
  moves: &LocationGraph,
  reg_colors: &HashMap<Reg, i32>,
  available_regs: &[Reg],
  used_callee_saved_regs: &mut IndexSet<Reg>,
) -> (Block, usize) {
  let var_colors = color_graph(conflicts, moves, locals, reg_colors);

  for (_, c) in &var_colors {
    if let Some(&reg) = available_regs.get(c.0 as usize) {
      if reg.is_callee_saved() {
        used_callee_saved_regs.insert(reg);
      }
    }
  }

  let block = Block {
    global: block.global,
    code: block
      .code
      .into_iter()
      .map(|instr| {
        assign_instr_var(instr, var_store, &var_colors, available_regs)
      })
      .collect(),
  };

  let num_locals = var_colors
    .values()
    .filter(|c| c.0 as usize >= reg_colors.len())
    .collect::<HashSet<_>>()
    .len();

  (block, num_locals)
}

fn assign_instr_var(
  instr: Instr<IdxVar>,
  var_store: &mut VarStore,
  var_colors: &IndexMap<Var, Color>,
  available_regs: &[Reg],
) -> Instr {
  let mut assign = |arg: Arg<IdxVar>| match arg {
    Arg::Deref(reg, i) => Arg::Deref(reg, i),
    Arg::Reg(reg) => Arg::Reg(reg),
    Arg::Imm(i) => Arg::Imm(i),
    Arg::Var(var) => {
      let i = var_colors[&var_store.get(var)].0 as usize;
      if let Some(&reg) = available_regs.get(i) {
        Arg::Reg(reg)
      } else {
        let local = i - available_regs.len() + 1;
        Arg::Deref(Reg::Rbp, -8 * local as i32)
      }
    }
    Arg::ByteReg(_) => unimplemented!(),
  };

  let instr = match instr {
    Instr::Add { src, dest } => Instr::Add {
      src: assign(src),
      dest: assign(dest),
    },
    Instr::Call(label, arity) => Instr::Call(label, arity),
    Instr::Jmp(label) => Instr::Jmp(label),
    Instr::Mov { src, dest } => Instr::Mov {
      src: assign(src),
      dest: assign(dest),
    },
    Instr::Neg(dest) => Instr::Neg(assign(dest)),
    Instr::Pop(dest) => Instr::Pop(assign(dest)),
    Instr::Push(src) => Instr::Push(assign(src)),
    Instr::Ret => Instr::Ret,
    Instr::Sub { src, dest } => Instr::Sub {
      src: assign(src),
      dest: assign(dest),
    },
    Instr::Syscall => Instr::Syscall,
    _ => unreachable!("{:?}", instr),
  };
  instr
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
struct Color(i32);

#[derive(Debug, Clone, PartialEq, Eq)]
struct NodeState {
  location: Location,
  color: Option<Color>,
  saturation: BTreeSet<i32>,
  num_moves: usize,
}

impl NodeState {
  fn new(location: Location, num_moves: usize) -> Self {
    Self {
      location,
      color: None,
      saturation: BTreeSet::new(),
      num_moves,
    }
  }
}

impl PartialOrd for NodeState {
  fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
    match self.saturation.len().partial_cmp(&other.saturation.len()) {
      Some(Ordering::Equal) => self.num_moves.partial_cmp(&other.num_moves),
      x => x,
    }
  }
}

impl Ord for NodeState {
  fn cmp(&self, other: &Self) -> Ordering {
    match self.saturation.len().cmp(&other.saturation.len()) {
      Ordering::Equal => self.num_moves.cmp(&other.num_moves),
      x => x,
    }
  }
}

/// * `graph` - interference graph
/// * `moves` - move graph
/// * `locals` - a slice of all local variables in this block
/// * `reg_colors` - a map of colors of all available registers
fn color_graph(
  graph: &LocationGraph,
  moves: &LocationGraph,
  locals: &[Var],
  reg_colors: &HashMap<Reg, i32>,
) -> IndexMap<Var, Color> {
  let mut queue = PriorityQueue::<NodeIndex, NodeState>::new();

  for node in graph.graph.node_indices() {
    let location = graph.graph[node];
    let move_node = moves.get_node(location);
    let num_moves =
      move_node.map_or(0, |node| moves.graph.neighbors(node).count());
    queue.push(node, NodeState::new(location, num_moves));
  }

  for node in graph.graph.node_indices() {
    if let Some(reg) = graph.graph[node].to_reg() {
      let color = reg_colors
        .get(&reg)
        .map_or_else(|| !reg.to_i32().unwrap(), |&c| c);
      for neighbor in graph.graph.neighbors(node) {
        queue.change_priority_by(&neighbor, |v| {
          v.saturation.insert(color);
        })
      }
      queue.remove(&node);
    }
  }

  let mut var_colors = IndexMap::new();

  while let Some((node, state)) = queue.pop() {
    let mut last_color_ix = -1;
    for &ix in &state.saturation {
      if ix < 0 {
        continue;
      }
      if ix == last_color_ix + 1 {
        last_color_ix = ix;
      } else {
        break;
      }
    }
    let mut color = Color(last_color_ix + 1);

    // move biasing
    let color_is_reg = (color.0 as usize) < reg_colors.len();
    for neighbor in moves.graph.neighbors(node) {
      if let Some(neighbor) = queue.get(&neighbor) {
        if let Some(neighbor_color) = neighbor.1.color {
          let ok = if color_is_reg {
            (neighbor_color.0 as usize) < reg_colors.len()
          } else {
            (neighbor_color.0 as usize) >= reg_colors.len()
          };
          if ok && !state.saturation.contains(&neighbor_color.0) {
            color = neighbor_color;
            break;
          }
        }
      }
    }

    for neighbor in graph.graph.neighbors(node) {
      queue.change_priority_by(&neighbor, |v| {
        v.saturation.insert(color.0);
      })
    }
    let var = state.location.to_var().unwrap();
    var_colors.insert(var, color);
  }

  for &var in locals {
    var_colors.entry(var).or_insert(Color(0));
  }

  var_colors
}

#[cfg(test)]
mod tests {
  use super::*;
  use crate::location_set::LocationSet;
  use asm::{Arg, Block, Label};
  use ch2::pass::select_instruction::Info as OldOldInfo;
  use indexmap::indexset;
  use insta::assert_snapshot;
  use maplit::hashmap;

  fn var(name: &str) -> Arg<IdxVar> {
    Arg::Var(IdxVar::new(name))
  }

  #[test]
  fn example_in_book() {
    use asm::Reg::*;
    use Arg::*;
    use Instr::*;
    let code = vec![
      Mov {
        src: Imm(1),
        dest: var("v"),
      },
      Mov {
        src: Imm(42),
        dest: var("w"),
      },
      Mov {
        src: var("v"),
        dest: var("x"),
      },
      Add {
        src: Imm(7),
        dest: var("x"),
      },
      Mov {
        src: var("x"),
        dest: var("y"),
      },
      Mov {
        src: var("x"),
        dest: var("z"),
      },
      Add {
        src: var("w"),
        dest: var("z"),
      },
      Mov {
        src: var("y"),
        dest: var("t"),
      },
      Neg(var("t")),
      Mov {
        src: var("z"),
        dest: Reg(Rax),
      },
      Add {
        src: var("t"),
        dest: Reg(Rax),
      },
      Jmp(Label::Conclusion),
    ];
    let label_live = hashmap! {
      Label::Conclusion => {
        let mut set = LocationSet::new();
        set.add_reg(Rax);
        set.add_reg(Rsp);
        set
      }
    };
    let prog = Program {
      info: OldOldInfo {
        locals: IndexSet::new(),
      },
      blocks: vec![(
        Label::Start,
        Block {
          global: false,
          code,
        },
      )],
    };
    let prog =
      super::super::liveness_analysis::analyze_liveness(prog, label_live);
    let prog = super::super::interference::build_interference(prog);
    let result = allocate_registers(prog, &[Rcx]);

    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn call() {
    use asm::Reg::*;
    use Arg::*;
    use Instr::*;
    let code = vec![
      Pop(Reg(Rdi)),
      Pop(Reg(Rsi)),
      Push(var("x")),
      Mov {
        src: Reg(Rbx),
        dest: var("w"),
      },
      Call("foo".to_owned(), 3),
      Add {
        src: Reg(Rax),
        dest: var("w"),
      },
      Jmp(Label::Conclusion)
    ];
    let label_live = hashmap! {
      Label::Conclusion => {
        let mut set = LocationSet::new();
        set.add_reg(Rax);
        set.add_reg(Rsp);
        set
      }
    };
    let prog = Program {
      info: OldOldInfo {
        locals: IndexSet::new(),
      },
      blocks: vec![(
        Label::Start,
        Block {
          global: false,
          code,
        },
      )],
    };
    let prog =
      super::super::liveness_analysis::analyze_liveness(prog, label_live);
    let prog = super::super::interference::build_interference(prog);
    let result = allocate_registers(prog, &[Rdx, Rdi, Rsi, R8, R12]);

    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn call_no_enough_registers() {
    use asm::Reg::*;
    use Arg::*;
    use Instr::*;
    let code = vec![
      Pop(Reg(Rdi)),
      Pop(Reg(Rsi)),
      Push(var("x")),
      Mov {
        src: Reg(Rbx),
        dest: var("w"),
      },
      Call("foo".to_owned(), 3),
      Add {
        src: Reg(Rax),
        dest: var("w"),
      },
      Jmp(Label::Conclusion)
    ];
    let label_live = hashmap! {
      Label::Conclusion => {
        let mut set = LocationSet::new();
        set.add_reg(Rax);
        set.add_reg(Rsp);
        set
      }
    };
    let prog = Program {
      info: OldOldInfo {
        locals: IndexSet::new(),
      },
      blocks: vec![(
        Label::Start,
        Block {
          global: false,
          code,
        },
      )],
    };
    let prog =
      super::super::liveness_analysis::analyze_liveness(prog, label_live);
    let prog = super::super::interference::build_interference(prog);
    let result = allocate_registers(prog, &[Rdx, Rdi, Rsi, R8]);

    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn mov_same_variables() {
    use asm::Reg::*;
    use Instr::*;
    let code = vec![
      Mov {
        src: var("x"),
        dest: var("t"),
      },
      Add {
        src: var("y"),
        dest: var("t"),
      },
      Mov {
        src: var("t"),
        dest: var("z"),
      },
      Add {
        src: var("w"),
        dest: var("z"),
      },
      Neg(var("x")),
      Neg(var("y")),
      Neg(var("z")),
      Neg(var("w")),
    ];
    let label_live = hashmap! {};
    let prog = Program {
      info: OldOldInfo {
        locals: indexset! {
          IdxVar::new("x"),
          IdxVar::new("y"),
          IdxVar::new("z"),
          IdxVar::new("t"),
          IdxVar::new("w"),
        },
      },
      blocks: vec![(
        Label::Start,
        Block {
          global: false,
          code,
        },
      )],
    };
    let prog =
      super::super::liveness_analysis::analyze_liveness(prog, label_live);
    let prog = super::super::interference::build_interference(prog);
    let result = allocate_registers(prog, &[Rdx, Rdi, Rsi, R8]);

    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn move_biasing_example_in_book() {
    use asm::Reg::*;
    use Arg::*;
    use Instr::*;
    let code = vec![
      Mov {
        src: Imm(1),
        dest: var("v"),
      },
      Mov {
        src: Imm(42),
        dest: var("w"),
      },
      Mov {
        src: var("v"),
        dest: var("x"),
      },
      Add {
        src: Imm(7),
        dest: var("x"),
      },
      Mov {
        src: var("x"),
        dest: var("y"),
      },
      Mov {
        src: var("x"),
        dest: var("z"),
      },
      Add {
        src: var("w"),
        dest: var("z"),
      },
      Mov {
        src: var("y"),
        dest: var("t"),
      },
      Neg(var("t")),
      Mov {
        src: var("z"),
        dest: Reg(Rax),
      },
      Add {
        src: var("t"),
        dest: Reg(Rax),
      },
      Jmp(Label::Conclusion),
    ];
    let label_live = hashmap! {
      Label::Conclusion => {
        let mut set = LocationSet::new();
        set.add_reg(Rax);
        set.add_reg(Rsp);
        set
      }
    };
    let prog = Program {
      info: OldOldInfo {
        locals: IndexSet::new(),
      },
      blocks: vec![(
        Label::Start,
        Block {
          global: false,
          code,
        },
      )],
    };
    let prog =
      super::super::liveness_analysis::analyze_liveness(prog, label_live);
    let prog = super::super::interference::build_interference(prog);
    let prog = super::super::move_biasing::build_move_graph(prog);
    let result = allocate_registers(prog, &[Rcx, Rdx, Rsi]);

    assert_snapshot!(result.to_string_pretty());
  }
}
