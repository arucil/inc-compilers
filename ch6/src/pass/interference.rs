#[cfg(test)]
mod tests {
  use asm::{Label, Program};
  use ast::{IdxVar, Type};
  use ch3::location_set::LocationSet;
  use ch4::pass::instruction_selection::Info as OldOldInfo;
  use ch4::pass::interference;
  use ch4::pass::liveness_analysis;
  use indexmap::indexmap;
  use insta::assert_snapshot;
  use maplit::hashmap;

  #[test]
  fn vector() {
    use asm::Reg::*;
    let blocks = asm::parse_blocks(
      |s| IdxVar::new(s),
      r#"
    start:
      pop rdi
      pop rsi
      push x
      mov w, rbx
      call rt_allocate, 2, gc
      add w, x
      jmp conclusion
    "#,
    );
    let label_live = hashmap! {
      Label::Conclusion => LocationSet::regs([Rax, Rsp])
    };
    let prog = Program {
      info: OldOldInfo {
        locals: indexmap! {
          IdxVar::new("x") => Type::Tuple(vec![Type::Int]),
          IdxVar::new("w") => Type::Int,
        },
      },
      blocks,
      ..Program::default()
    };
    let prog = liveness_analysis::analyze_liveness(prog, label_live);
    let result = interference::build_interference(prog);

    assert_snapshot!(format!("{}", result.info));
  }

  #[test]
  fn str() {
    use asm::Reg::*;
    let blocks = asm::parse_blocks(
      |s| IdxVar::new(s),
      r#"
    start:
      pop rdi
      pop rsi
      push x
      mov w, rbx
      call rt_allocate, 2, gc
      add w, x
      jmp conclusion
    "#,
    );
    let label_live = hashmap! {
      Label::Conclusion => LocationSet::regs([Rax, Rsp])
    };
    let prog = Program {
      info: OldOldInfo {
        locals: indexmap! {
          IdxVar::new("x") => Type::Str,
          IdxVar::new("w") => Type::Int,
        },
      },
      blocks,
      ..Program::default()
    };
    let prog = liveness_analysis::analyze_liveness(prog, label_live);
    let result = interference::build_interference(prog);

    assert_snapshot!(format!("{}", result.info));
  }
}
