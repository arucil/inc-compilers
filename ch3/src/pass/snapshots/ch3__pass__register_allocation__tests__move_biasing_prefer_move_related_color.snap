---
source: ch3/src/pass/register_allocation.rs
expression: result
---
{
    Var(
        2,
    ): Color(
        0,
    ),
    Var(
        0,
    ): Color(
        1,
    ),
    Var(
        1,
    ): Color(
        1,
    ),
    Var(
        3,
    ): Color(
        0,
    ),
    Var(
        4,
    ): Color(
        1,
    ),
}
